src/v/kafka/requests/consumer_records.h

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "kafka/types.h"
+
+#include <optional>
+
+namespace kafka {
+
+///\brief consumer_records are a concatenation of multiple model::record_batch
+/// on the wire; without a size header (c.f. a kafka array<thing>)
+class consumer_records {
+public:
+    consumer_records() = default;
+
+    explicit consumer_records(std::optional<iobuf> buf)
+      : _record_set(std::move(buf)) {}
+
+    // Returns true if the record_set is nullopt or empty
+    bool empty() const { return !_record_set || _record_set->empty(); }
+
+    // Returns the size of the buffer, or -1 if it's nullopt
+    size_t size_bytes() const {
+        return _record_set ? _record_set->size_bytes() : -1;
+    }
+
+    // Check that the iobuf isn't nullopt
+    constexpr explicit operator bool() const noexcept {
+        return _record_set.operator bool();
+    }
+
+    // Release any remaining iobuf that hasn't been consumed
+    std::optional<iobuf> release() && { return std::move(_record_set); }
+
+private:
+    std::optional<iobuf> _record_set;
```

Comment:
Can we simplify `consumer_records` by removing the optional and deal with nullopt at each call site? For example on the Kafka side I think that when we read the optional iobuf off the wire we could probably handle it right then (which I believe is to return an error).

Seems like optional here makes empty/size_bytes etc... some what awkward in an otherwise really useful utility. Would that make sense on the proxy side?

Reply:
Sounds good to me. I don't really like the subtleness of having to check `operator bool()`, `empty()`, `size_bytes()`.

Reply:
right, it just pushes out edge cases to things like size_bytes being -1 or asserting non-empty in there.

Reply:
It looks like fetch_request.cc always constructs with an empty iobuf - I'll give it a go.